README.md
# PaapiMini

* For Japanese [README_ja.md](./README_ja.md)
* This is a Gem for easy Amazon product search.<br>
Before using it, you need to pass Amazon Associate review and get ACCESS_KEY etc. as preparation.
* [Amazon Product Advertising API 5.0](https://webservices.amazon.com/paapi5/documentation/)　is supported.
* This Gem is implemented using only the Ruby standard-attached library, so it does not depend on other Gems.

* Ruby versions that have been tested by rbenv
```
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin20]
ruby 3.0.0rc1 (2020-12-20 master 8680ae9cbd) [x86_64-darwin20]
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [x86_64-darwin21]
```
* We have not tested Ruby for Windows at this time.<br>
Also, since the author's Amazon Associate review is conducted by Amazon.co.jp, we are only able to test from this region.<br>
Therefore, the default behavior of the API is Amazon.co.jp. It is possible to change it with an argument.<br>
We would appreciate it if you could provide us with feedback if you can confirm that it works on the Windows version or in regions other than Amazon.co.jp.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'paapi_mini'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install paapi_mini

## Usage

```ruby
    require 'paapi_mini'

    # The keyword argument to the search_items method can be used to refine the condition.
    # Please refer to the official reference for the specifications of each argument.
    # See Also https://webservices.amazon.com/paapi5/documentation/search-items.html
    # Default Values.
    # item_page: 1,
    # min_price: 1,
    # max_price: 100000,
    # min_review_rating: 1,
    # host: "webservices.amazon.co.jp",
    # region: "us-west-2",
    # market_place: "www.amazon.co.jp"
    
    res = PaapiMini::search_items(ENV['ACCESS_KEY'], ENV['SECRET_KEY'], ENV['PARTNER_TAG'],
                                  keywords: "Ruby", min_review_rating: 3)

    # res is Net::HTTPResponse Object
    # See Also https://docs.ruby-lang.org/
    puts res.code
    puts res.body
```

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## If you like this Gem, please help us!

* Development is currently suspended. This is because the API response results in an error if no Amazon Associate sales results are generated for 30 days.
* If you like this Gem, we would appreciate it if you could buy Amazon products from [this link](https://amzn.to/3Ta7pvV).
* It doesn't matter what the price is, it doesn't matter if it is another product that you followed from this link.
* Once we have achieved sales and the API restrictions are released, we would like to implement other API support such as get_items, etc.

CHANGELOG.md
# CHANGELOG

## 0.1.1, 0.1.2
- document update only.

## 0.1.0
- Added. ChangeLog And Release.

test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "paapi_mini"

require "minitest/autorun"


test/test_paapi_mini.rb
# frozen_string_literal: true

require_relative "test_helper"
require_relative '../lib/paapi_mini'

class TestPaapiMini < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::PaapiMini::VERSION
  end

  def test_it_does_something_useful
    res = PaapiMini::search_items(ENV['ACCESS_KEY'], ENV['SECRET_KEY'], ENV['PARTNER_TAG'], "Ruby")
    status_code = res.code.to_i
    puts status_code
    puts res.body
    assert status_code != 404
  end
end


lib/paapi_mini.rb
# frozen_string_literal: true

require_relative "paapi_mini/version"
require_relative 'aws_v4_auth'
require_relative 'http_client'

module PaapiMini
  class Error < StandardError; end

  def self.search_items(access_key,
                       secret_key,
                       partner_tag,
                       keywords,
                       item_page: 1,
                       min_price: 1,
                       max_price: 100000,
                       min_review_rating: 1,
                       host: "webservices.amazon.co.jp",
                       region: "us-west-2",
                       market_place: "www.amazon.co.jp")

    aws_auth = Auth.create_search_items(access_key,
                                        secret_key,
                                        partner_tag,
                                        keywords: keywords,
                                        item_page: item_page,
                                        min_price: min_price,
                                        max_price: max_price,
                                        min_review_rating: min_review_rating,
                                        host: host,
                                        region: region,
                                        market_place: market_place)

    aws_auth_headers = aws_auth.make_headers
    aws_auth_headers['Content-Type'] = 'application/json; charset=utf-8'

    return Http.post(aws_auth.make_uri, aws_auth.payload, aws_auth_headers)
  end


end


lib/http_client.rb
require 'net/http'
require 'net/https'
require 'uri'

module PaapiMini
module Http

  # Exception
  class Error < StandardError
    def initialize(msg = 'HttpClient Error Message.')
      super
    end
  end

  # Post
  def self.post(uri, payload, headers, options = {})
    begin
      parsed_uri, http = create_http(uri, options.has_key?(:proxy_url) ? options[:proxy_url] : nil)
      req = Net::HTTP::Post.new(parsed_uri.request_uri)

      headers.each do |key, value|
        req[key.to_s] = value.to_s
      end

      if options.has_key?(:basic_auth)
        req.basic_auth(options[:basic_auth][0], options[:basic_auth][1])
      end

      req.body = payload
      http.request(req)
    rescue => e
      raise PaapiMini::Http::Error.new("#{e.message} #{e.backtrace}" )
    end
  end

  class << self
    private

    def create_http(uri, proxy_uri)
      parsed_uri = URI.parse(uri)
      if parsed_uri.to_s == ""
        raise PaapiMini::Http::Error.new("URI parse error. uri=>#{uri}")
      end

      http = if !proxy_uri.nil? && proxy_uri.length > 0
               parsed_proxy_uri = URI.parse(proxy_uri)
               if parsed_proxy_uri.to_s == ""
                 raise PaapiMini::Http::Error.new("URI parse error. proxy_uri=>#{proxy_uri}")
               end

               # Proxy Server
               proxy_class = Net::HTTP::Proxy(parsed_proxy_uri.host.to_s, parsed_proxy_uri.port)
               proxy_class.new(parsed_uri.host, parsed_uri.port)
             else
               Net::HTTP.new(parsed_uri.host.to_s, parsed_uri.port)
             end

      http.open_timeout = 5
      http.read_timeout = 5

      if parsed_uri.scheme == 'https'
        http.use_ssl     = true
        http.verify_mode = OpenSSL::SSL::VERIFY_PEER
      end
      [parsed_uri, http]
    end
  end
end
end

lib/aws_v4_auth.rb
require 'digest'
require 'openssl'
require 'time'
require 'uri'

module PaapiMini

  class Auth

    attr_reader :access_key,
                :secret_key,
                :path,
                :region,
                :service,
                :http_method_name,
                :headers,
                :payload,
                :xamz_date,
                :current_date,
                :signed_headers

    HMAC_ALGORITHM = 'AWS4-HMAC-SHA256'
    AWS4_REQUEST = 'aws4_request'

    public

    # Make Auth Headers Strings for Amazon Product Advertising API 5.0
    # reference: https://webservices.amazon.com/paapi5/documentation/without-sdk.html
    #            https://webservices.amazon.com/paapi5/documentation/sending-request.html#signing
    # important! https://webservices.amazon.co.jp/paapi5/scratchpad/
    def make_headers
      @headers['X-Amz-Date'] = @xamz_date
      @headers['X-Amz-Content-Sha256'] = OpenSSL::Digest::SHA256.hexdigest(@payload)

      canonical_url = prepare_canonical_request(@http_method_name, @path, @headers['X-Amz-Content-Sha256'])
      string_to_sign = prepare_string_to_sign(canonical_url)
      signature = calculate_signature(string_to_sign)
      @headers['Authorization'] = build_authorization_string(signature)
      Hash[@headers]
    end

    # MakeURI
    # this call after method create_search_items required
    def make_uri
      "https://#{@headers['Host']}#{@path}"
    end

    def prepare_canonical_request(http_method_name, path, payload)
      canonical_url = []
      uri = URI.parse(path.to_s)
      canonical_url << http_method_name
      canonical_url << "#{uri.path}\n"

      @signed_headers = []
      headers = @headers.sort_by(&:first)
      canonical_headers = []
      headers.each do |key, value|
        @signed_headers << key.downcase
        canonical_headers << "#{key.downcase}:#{value}"
      end

      canonical_url << "#{canonical_headers.join("\n")}\n"
      canonical_url << @signed_headers.join(";")
      canonical_url << payload
      canonical_url.join("\n")
    end

    def prepare_string_to_sign(canonical_url)
      string_to_sign = []
      string_to_sign << HMAC_ALGORITHM
      string_to_sign << @xamz_date
      string_to_sign << "#{@current_date}/#{@region}/#{@service}/#{AWS4_REQUEST}"
      string_to_sign << OpenSSL::Digest::SHA256.hexdigest(canonical_url)
      string_to_sign.join("\n")
    end

    def calculate_signature(string_to_sign)
      OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha256'), signature_key(), string_to_sign)
    end

    def build_authorization_string(signature)
      build_authorization_string = []
      build_authorization_string << "Credential=#{credentials_path}"
      build_authorization_string << "SignedHeaders=#{@signed_headers.join(";")}"
      build_authorization_string << "Signature=#{signature}"
      "#{HMAC_ALGORITHM} #{build_authorization_string.join(", ")}"
    end

    def credentials_path
      credentials = [@access_key, @current_date, region, service, AWS4_REQUEST]
      credentials.join("/")
    end

    protected


    def initialize(access_key, secret_key, path, region, service, http_method_name, headers, payload)
      @access_key, @secret_key, @path = access_key, secret_key, path
      @region, @service, @http_method_name = region, service, http_method_name
      @headers, @payload = headers, payload
      @xamz_date, @current_date = Auth::utc_timestamp(Time.now)
    end



    # https://docs.aws.amazon.com/general/latest/gr/signature-v4-examples.html#signature-v4-examples-ruby
    def signature_key
      secret = "AWS4#{@secret_key}"
      date = Auth::hmac_sha_256(secret, @current_date)
      region = Auth::hmac_sha_256(date, @region)
      service = Auth::hmac_sha_256(region, @service)
      Auth::hmac_sha_256(service, AWS4_REQUEST)
    end

    class << self

      # SearchItems
      # https://webservices.amazon.com/paapi5/documentation/search-items.html
      # SearchIndex
      # https://webservices.amazon.com/paapi5/documentation/locale-reference/japan.html
      def create_search_items(access_key,
                              secret_key,
                              partner_tag,
                              keywords:,
                              item_page: 1,
                              min_price: 1,
                              max_price: 100000,
                              min_review_rating: 1,
                              host: "webservices.amazon.co.jp",
                              region: "us-west-2",
                              market_place: "www.amazon.co.jp")

        service = 'ProductAdvertisingAPI'
        http_method_name = 'POST'
        api_path = '/paapi5/searchitems'

        headers = {}
        headers['Content-Encoding'] = 'amz-1.0'
        headers['Host'] = host
        headers['X-Amz-Target'] = 'com.amazon.paapi5.v1.ProductAdvertisingAPIv1.SearchItems'

        request_payload = <<~EOS
          {
            "Keywords": "#{keywords}",
            "Resources": [
              "BrowseNodeInfo.WebsiteSalesRank",
              "CustomerReviews.Count",
              "CustomerReviews.StarRating",
              "Images.Primary.Small",
              "Images.Primary.Medium",
              "ItemInfo.ByLineInfo",
              "ItemInfo.ContentInfo",
              "ItemInfo.ContentRating",
              "ItemInfo.Classifications",
              "ItemInfo.ManufactureInfo",
              "ItemInfo.ProductInfo",
              "ItemInfo.TechnicalInfo",
              "ItemInfo.Title",
              "ItemInfo.TradeInInfo",
              "Offers.Listings.Condition",
              "Offers.Listings.Condition.ConditionNote",
              "Offers.Listings.Condition.SubCondition",
              "Offers.Listings.DeliveryInfo.IsAmazonFulfilled",
              "Offers.Listings.DeliveryInfo.IsFreeShippingEligible",
              "Offers.Listings.DeliveryInfo.IsPrimeEligible",
              "Offers.Listings.DeliveryInfo.ShippingCharges",
              "Offers.Listings.IsBuyBoxWinner",
              "Offers.Listings.LoyaltyPoints.Points",
              "Offers.Listings.MerchantInfo",
              "Offers.Listings.Price",
              "Offers.Listings.ProgramEligibility.IsPrimeExclusive",
              "Offers.Listings.ProgramEligibility.IsPrimePantry",
              "Offers.Listings.Promotions",
              "Offers.Listings.SavingBasis",
              "Offers.Summaries.HighestPrice",
              "Offers.Summaries.LowestPrice",
              "Offers.Summaries.OfferCount",
              "ParentASIN",
              "SearchRefinements"
            ],
            "SearchIndex": "All",
            "PartnerTag": "#{partner_tag}",
            "PartnerType": "Associates",
            "Marketplace": "#{market_place}",
            "Operation": "SearchItems",
            "ItemPage": #{item_page},
            "MinReviewsRating": #{min_review_rating},
            "MinPrice": #{min_price},
            "MaxPrice": #{max_price},
            "CurrencyOfPreference":"JPY",
            "Condition":"New",
            "DeliveryFlags":["FulfilledByAmazon"]
          }
        EOS

        Auth.new(access_key, secret_key, api_path, region, service, http_method_name, headers, request_payload)
      end

      def hmac_sha_256(key, base_string)
        OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha256'), key, base_string)
      end

      def utc_timestamp(time)
        now = time.utc
        return now.strftime("%Y%m%dT%H%M%SZ"), now.strftime("%Y%m%d")
      end

    end

  end
end
